WebGPURenderer: Support MSAA with Postprocessing

RenaudRohlinger opened the pull request:
**Description**

The WebGPU did not support MSAA with a post-processing pipeline. This PR fixes the issue and adds a way to specify options, such as `samples`, to `PassNode`.


*This contribution is funded by [Utsubo](https://utsubo.com)*

RenaudRohlinger committed:
WebGPURenderer: Support MSAA with Postprocessing

RenaudRohlinger committed:
cleanup

Mugen87 commented on examples/webgpu_postprocessing_3dlut.html:
```diff
@@ -134,7 +135,7 @@
 				postProcessing = new THREE.PostProcessing( renderer );
 
-				const scenePass = pass( scene, camera );
+				const scenePass = pass( scene, camera, { samples: params.samples } );
```
Do you mind picking another example which does not use a HDR workflow?

MSAA is not a good choice for HDR since depending on the sampled HDR texel value, the anti-aliasing effect get more or less disabled, see https://github.com/mrdoob/three.js/issues/26954#issuecomment-1759721712.

FX post-processing like FXAA or SMAA or the more high-quality SSAA/TAA are more appropriate for such scenes. We can add them step by step in the upcoming months. 

Even if the quality isn't that awesome, I would recommend to start with FXAA since it's always nice to have this performant technique as a option. `FXAAShader` can maybe be converted with TSL transpiler.

RenaudRohlinger replied:
Would `webgpu_loader_gltf.html` work then?

RenaudRohlinger replied:
I removed the example for now, maybe we can include it in the future into another example, I feel like `misc_controls_fly` would not really highlight the difference between the different amount of `samples` and creating a new example just for this seems too much.

sunag replied:
I think we could honor the samplers definied in `Renderer` for `pass()`, in case of `options=null`, with this logic `WebGPURenderer( { samples: params.samples } )` should work for all pass()

RenaudRohlinger replied:
Not sure if that's what you mean but basically check if the renderer is antialiased in `setup` of `PassNode` and if `this.options.samples === undefined` use `samples = 4` instead of `0`?

sunag replied:
Exactly, I think we could store samples in Renderer and reuse it in `setup( builder ) { ... builder.renderer }` moment to update the RenderTarget textures.

RenaudRohlinger committed:
cleanup and revert 3dlut

RenaudRohlinger committed:
refactor AA and samples logic

RenaudRohlinger committed:
Merge remote-tracking branch 'upstream/dev' into utsubo/feat/postprocess-msaa

RenaudRohlinger committed:
add multisampled_texture support

RenaudRohlinger committed:
cleanup

RenaudRohlinger committed:
more cleanup

Mugen87 commented on src/nodes/display/PassNode.js:
```diff
@@ -130,7 +131,18 @@ class PassNode extends TempNode {
+		for ( let i = 0; i < this.renderTarget.textures.length; i ++ ) {
+
+			const type = this.renderTarget.textures[ i ].type;
+			this.renderTarget.textures[ i ].isMultisampleRenderTargetTexture = this.renderTarget.samples > 1 && type !== HalfFloatType && type !== FloatType;
```
What is the purpose of `type !== HalfFloatType && type !== FloatType`? 

Does that mean multisampled (half) float textures are not allowed?

RenaudRohlinger committed:
temporary patch webgl backend

RenaudRohlinger commented:
I cleaned up and refactored the samples and antialiasing logic. Previously, it was under `backend.parameters.antialias` and was only handled by the WebGPUBackend.
I moved the antialiasing settings to the Renderer and introduced a `renderer.samples` property. If this new property is set to a value greater than 0, it will override the antialias logic. The `antialias` property now simply acts as a shortcut for setting `renderer.samples` to 4, allowing developers to initialize the sample count directly from the Renderer, for example, by using `new WebGPURenderer({ samples: 2 })`.

Since WebGPU only supports sample counts of 1 and 4, I refactored the `WebGPUUtils.getSampleCount` method and used it everywhere it's relevant to ensure the sample count is always translated appropriately in that backend. For the WebGLBackend, it will use the sample count defined in the Renderer, which supports any sample amount, usually with a maximum of 8.

Here is most of the refactor:
https://github.com/mrdoob/three.js/pull/28784/commits/811fd9a3f9c687164dfca59bcbef445c06fdb557

Following this refactor, I discovered that in the case of multisampled textures in WebGPU, we were always actually using two textures: one extra that resolves the multisampled one. Based on the specs, it seems that WebGPU supports direct rendering of multisampled 2D and depth 2D textures for non-float textures ( 👉 https://www.w3.org/TR/webgpu/#dom-gputexturebindinglayout-multisampled).

Therefore, I added multisample texture support to the WebGPUBackend and updated the WGSLNodeBuilder to support both types: `texture_multisampled_2d` and `texture_depth_multisampled_2d`.

Small notes:

- The multisampling process for textures will continue to work as before if the type is Float or HalfFloat. This is because WebGPU does not support directly rendering to multisampled textures with these types. ["unfilterable-float"](https://www.w3.org/TR/webgpu/#dom-gputexturesampletype-unfilterable-float) seems to silent the issue but the antialiasing doesn't seems to get performed in that case. (To answer your comment, /cc @Mugen87)
- I have not implemented this feature in the WebGLBackend. Given that this is already a substantial PR, I prefer to address this support in a separate, future PR.

Aside from this, while testing most of the examples, I noticed that `webgpu_postprocessing_anamorphic.html`, `webgpu_compute_particles_snow.html`, and `webgpu_custom_fog_background` all look different from the examples in production of r166, and this issue is also present in dev.
/cc @sunag @mrdoob @Mugen87

RenaudRohlinger committed:
fix

Mugen87 commented:
> (To answer your comment, also I removed that block

It seems the check is still present though.

Um, the logic that configures `isMultisampleRenderTargetTexture` is something that should not belong to `PassNode`, imo. How about moving this into the renderer?

`isMultisampleRenderTargetTexture` is currently only configured by `PassNode` but it seems that should be done no matter of a render target was configured.

Mugen87 commented:
> I have not implemented this feature in the WebGLBackend. Given that this is already a substantial PR, I prefer to address this support in a separate, future PR.

I don't think it makes sense to do that in the near future since you need the `WEBGL_multisampled_render_to_texture` which is only supported by the Oculus Browser so far. And since we can't use XR with `WebGPURenderer` right now there is no possibility to test this.

Mugen87 commented:
Um, given that post processing in HDR requires half float and `THREE.HalfFloatType` is the default right niw, I doubt that rendering multisampled to a color renderable texture is that important right now for `PassNode`.

I would probably split the PR and support for `texture_multisampled_2d` and `texture_depth_multisampled_2d` with a different PR (independently of `PassNode`).

RenaudRohlinger committed:
feedbacks

RenaudRohlinger committed:
resolveTarget doesn't exist for depth textures

RenaudRohlinger commented:
The `texture_depth_multisampled_2d` is necessary to support MSAA with PassNode because GPURenderPassDepthStencilAttachment doesn't support using a `resolveTarget`:
![image](https://github.com/mrdoob/three.js/assets/15867665/93c74072-807e-4055-aebc-147f2573371f)
Which means that the depth texture associated with that RenderPass will be multisampled if the ColorBuffer is multisampled.
Unlike the color that is resolved to a normal texture target, the depth will be sent to the GPU as a multisampled texture, which means WGSL should read it as `texture_depth_multisampled_2d` in order to work and render something. That's why I had to implement this support in WGSL.

Also, I cleaned up `PassNode` and removed the multisampling for the color passes, which is now reduced to a single line for the `isMultisampleRenderTargetTexture` flag. Unfortunately, I have to set the `depthTexture.isMultisampleRenderTargetTexture` flag in the `PassNode.setup` function if we want to fallback to the renderer.samples, as the texture creation seems to occur after the `WGSLNodeBuilder` code runs. This would otherwise generate incorrect WGSL for the not-yet-updated texture.

Mugen87 commented:
Thanks for the additional information!

> Unfortunately, I have to set the depthTexture.isMultisampleRenderTargetTexture flag in the PassNode.setup function if we want to fallback to the renderer.samples, as the texture creation seems to occur after the WGSLNodeBuilder code runs. This would otherwise generate incorrect WGSL for the not-yet-updated texture.

Maybe @sunag has an idea how to hide this bit. Ideally, we evaluate whether a depth texture is multisampled or not inside the renderer and not in components using the renderer.

Mugen87 commented on src/nodes/display/PassNode.js:
```diff
@@ -130,7 +131,18 @@ class PassNode extends TempNode {
+		this.renderTarget.samples = this.options.samples === undefined ? renderer.samples : this.options.samples;
+
+		// Disable MSAA for WebGL backend for now
+		if ( renderer.backend.isWebGLBackend === true ) {
```
Is this exception still required? My understand is you have added this bit since rendering multisampled to a color renderable texture does not yet work in the WebGL backend. But can at least the "normal" MSAA be used?

RenaudRohlinger replied:
Unfortunately yes it's still needed. I noticed that using PassNode with `samples > 0` doesn't seem to render anything in WebGL. I believe this might be due to the need for an additional frameBuffer somewhere in the render pipeline, which I couldn't easily resolve. That's why I added this comment.

sunag replied:
Could we add this in `WebGLBackend.beginRender()` for now?

RenaudRohlinger replied:
I wouldn't recommend so because basic `THREE.RenderTarget` setups works just fine `with samples > 0`. For example `webgpu_multisampled_renderbuffers.html`.

sunag replied:
Let's merge this PR then! It's possible do it in another PR, currently already a step forward in that sense. So the future goal is for us to move this and `isMultisampleRenderTargetTexture` into the renderer as @Mugen87 said, when this is ready examples like `webgpu_backdrop_water` should work too.
